File: boj/tree/15681.py
'''
    https://www.acmicpc.net/problem/15681
    (tree and query)

    reference
        https://www.acmicpc.net/source/69400180
        (ckswofl, 404 ms)
        https://www.acmicpc.net/source/87297443
        (vincent0116, 364 ms)

'''

import sys

NUM_VERTEX = 100_000    # 10^5
NUM_QUERY = 100_000     # 10^5

N, R, Q = list(map(int, input().split()))

graph = [[] for _ in range(N + 1)]
visited = [False] * (N + 1)
dp = [0] * (N + 1)

# sys.setrecursionlimit(NUM_VERTEX + 1)       # recursion error
sys.setrecursionlimit(N << 1)      # AC

def dfs(curr : int) -> int:
    visited[curr] = True

    cnt = 1     # count itself
    for next in graph[curr]:
        if not visited[next]:
            cnt += dfs(next)

    dp[curr] = cnt

    return cnt

def solve():
    dfs(R)

for _ in range(N - 1):
    # U, V = list(map(int, input().split()))                # TLE
    # U, V = list(map(int, sys.stdin.readline().split()))
    U, V = map(int, sys.stdin.readline().split())
    graph[U].append(V)
    graph[V].append(U)

solve()

for _ in range(Q):
    # U = list(map(int, input().split()))[0]
    # U = int(input())                                      # TLE
    U = int(sys.stdin.readline())

    print(dp[U])

File: boj/tree/15681.cc
/*
    https://www.acmicpc.net/problem/15681
    (tree and query)
*/

#include <iostream>
#include <vector>

#define ll long long
#define endl '\n'

const int NUM_VERTEX = 100'000;     // 10^5
const int NUM_QUERY = 100'000;      // 10^5


int N, R, Q;
int U, V;

std::vector<int> graph[NUM_VERTEX + 1];
bool visited[NUM_VERTEX + 1];
int dp[NUM_VERTEX + 1];

int dfs(int curr){
    visited[curr] = true;

    int cnt = 1;
    for(int next : graph[curr]){
        if(!visited[next]){
            cnt += dfs(next);
        }
    }

    return dp[curr] = cnt;
}

void solve(){
    dfs(R);
}

void input(){
    std::cin >> N >> R >> Q;
    for(int ith=1; ith<=N-1; ++ith){
        std::cin >> U >> V;
        graph[U].push_back(V);
        graph[V].push_back(U);
    }

    solve();

    for(int ith=1; ith<=Q; ++ith){
        std::cin >> U;
        std::cout << dp[U] << endl;
    }
}

int main(){

    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    input();

    solve();

    return 0;
}
